src/translator/quality_estimator.cpp

```diff
@@ -0,0 +1,133 @@
+            break;
+        }
+      current_str.erase(0, pos + delimiter.length());
+      line_position++;
+    }
+    // last line
+   QualityEstimator::initVector(this->intercept, current_str);
+}
+
+void QualityEstimator::initVector(std::vector<float> &emptyVector, std::string line){
+    std::istringstream ss(line);
+  
+    std::string elem;
+    while (ss >> elem) 
+    {
+        emptyVector.push_back(std::stof(elem));
+    }
+}
+
+void QualityEstimator::mapBPEToWords(Response& sentence, Words words){
+    //The result is a vector of maps
+    // each element of vector is a map containing the bpes of a word
+    // each map key is a tuple consisting of the begining and end of the specific subword
+    std::vector<std::map<std::tuple<size_t, size_t>, float>> target_words;
+    for (int sentenceIdx = 0; sentenceIdx < sentence.size(); sentenceIdx++) {
+        auto &quality = sentence.qualityScores[sentenceIdx];
+        size_t wordIdx = 0;
+        bool first = true;
+        std::map<std::tuple<size_t, size_t>, float> map;
+        float previous_p;
+        for (auto &p : quality.word) {
+            if (words[wordIdx].toString()==words[wordIdx].DEFAULT_EOS_ID.toString()){
+                break;
+            }
+            ByteRange subword = sentence.target.wordAsByteRange(sentenceIdx, wordIdx);
+            size_t subword_begin = subword.begin;
+            size_t subword_end = subword.end;
+            char first_word = sentence.target.text[subword_begin];
+            wordIdx++;
+            if (first){
+                first = false;
+                previous_p = p;
+                map[std::make_tuple(subword_begin, subword_end)] = previous_p;
+                continue;
+            }
+            if (first_word == ' '){
+                target_words.push_back(map);
+                map.clear();
+                map[std::make_tuple(subword_begin+1, subword_end)] = p;
+            }
+            else{
+                map[std::make_tuple(subword_begin, subword_end)] = p;
+            }
+        }
+        //last word
+        if (!map.empty()){
+            target_words.push_back(map);
+        }
+    }
+    this->mapping = target_words;
+}
+ intgemm::AlignedVector<QualityEstimator::FeatureVector> QualityEstimator::extractFeatures(Response& sentence){
+     auto quality_scores = sentence.qualityScores[0]; //this is expected to be only one
+     auto map = this->mapping;
+     const intgemm::Index n_rows = map.size();
+     const intgemm::Index width = 64;
+     intgemm::AlignedVector<QualityEstimator::FeatureVector> feature_matrix(n_rows * width);
+     intgemm::Index current_idx = 0;
+     for(auto &word: map) {
+        bool first=true;
+        size_t first_subword_idx;
+        size_t last_subword_idx;
+        float bpe_scores_sum = 0;
+        float bpe_scores_min;
+        for (auto &subword:word){
+            auto tmp = subword.first;
+            float bpe_score = subword.second;
+            if (first){
```

Comment:
Anti-pattern of branch inside a for loop to check if it's first instead of putting code before for loop.  Could have initialized to infinite to avoid a special case here.